### PR TITLE
Fix multiple maintenance call triggered by patch call for status update

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area control-plane
/kind bug

**What this PR does / why we need it**:
Previously if shoot maintenance is triggered by  `gardener.cloud/operation`: `maintain` annotation then it can cause multiple maintenance call cause here https://github.com/gardener/gardener/blob/afa6db45466a5f1c171a3848ef0f86168464d92c/pkg/controllermanager/controller/shoot/maintenance/reconciler.go#L194-L197 
patch call will trigger another reconciliation and it will satisfy the required predicate of the controller.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Before requesting reviews, it would be great if the PR description is properly maintained.

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go [hunk 1]
```diff
@@ -274,36 +301,41 @@ func maintainTasks(shoot *gardencorev1beta1.Shoot, config config.ShootMaintenanc
 }
 
 // MaintainMachineImages updates the machine images of a Shoot's worker pools if necessary
-func MaintainMachineImages(log logr.Logger, shoot *gardencorev1beta1.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) ([]string, error) {
+func MaintainMachineImages(log logr.Logger, shoot *gardencorev1beta1.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) ([]string, func(*gardencorev1beta1.Shoot), error) {
 	var reasonsForUpdate []string
+	workers := shoot.Spec.Provider.Workers
 
-	for i, worker := range shoot.Spec.Provider.Workers {
+	for i, worker := range workers {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/maintenance/reconciler.go]: Can you revert this? I don't think this works and there is no benefit of having this `workers` variable either.

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go [hunk 2]
```diff
@@ -186,16 +200,29 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
...
 	}
 
-	if err := r.Client.Status().Patch(ctx, newShoot, patch); err != nil {
+	// here `gardener.cloud/operation`: `maintain` is removed so when patch call triggers
+	// another reconciliation it should not be picked by the controller
+	if hasMaintainNowAnnotation(shoot) {
+		delete(shoot.Annotations, v1beta1constants.GardenerOperation)
+	}
+
+	if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
```
Comment (plkokanov) [inline pkg/controllermanager/controller/shoot/maintenance/reconciler.go:202]: What if the patch status fails and the controller is restarted? Or even if the patch succeeds but the controller is restarted before the maintenance operation is completely finished? Then the maintenance operation will not be retried.

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go (hunk 2 above)
Comment (acumino) [reply pkg/controllermanager/controller/shoot/maintenance/reconciler.go:202]: In the first case if the patch call fails and the controller is restarted then the maintenance operation will be retried cause shoot will still have the required maintain operation annotation, but in the second case if the patch passes and controller restarts before the update call then maintenance operation will not be retried. I will open a revert PR for https://github.com/gardener/gardener/pull/6725 to not delay the release process.
/close

**Pull request closed without merging**